scanner/fetcher.go

```diff
@@ -137,9 +138,9 @@ func (f *Fetcher) Run(ctx context.Context, fn func(EntryBatch)) error {
 }
 
 // genRanges returns a channel of ranges to fetch, and starts a goroutine that
-// sends things down this channel. The goroutine terminates when all ranges
-// have been generated, or if context is cancelled.
-func (f *Fetcher) genRanges(ctx context.Context) <-chan fetchRange {
+// sends things down this channel. The goroutine terminates if all ranges have
+// been generated, Stopper is Done, or context is cancelled.
```

Comment:
I still don't think it makes sense to have two different mechanisms to stop a `Fetcher`, one standard and one non-standard.

If you have two chunks of code that you want to stop at slightly different times, maybe re-arrange things so you can use two distinct contexts.

Reply:
I think in the `Fetcher` case it's doable by separating the workers pool from scheduler. We can cancel scheduler immediately so that it stops producing new jobs, but the pool will continue doing the started tasks until its context is canceled after a grace period.

Not so sure about `Controller` though. Maybe we could have a `Shutdown` method similar to [http.Server.Shutdown](https://golang.org/pkg/net/http/#Server.Shutdown).

I was in the process of factoring out workers pool anyway (in order to share the same pool between multiple "tenants", and also to re-use its code in multiple places), so I'll postpone this PR until I do that.